src/wasm/wasm-binary.cpp

```diff
@@ -516,38 +522,61 @@ uint32_t WasmBinaryWriter::getTypeIndex(HeapType type) const {
   return it->second;
 }
 
-void WasmBinaryWriter::writeFunctionTableDeclaration() {
-  if (!wasm->table.exists || wasm->table.imported()) {
+void WasmBinaryWriter::writeTableDeclarations() {
+  if (importInfo->getNumDefinedTables() == 0) {
+    // std::cerr << std::endl << "(WasmBinaryWriter::writeTableDeclarations) No
+    // defined tables found. skipping" << std::endl;
     return;
   }
-  BYN_TRACE("== writeFunctionTableDeclaration\n");
+  BYN_TRACE("== writeTableDeclarations\n");
   auto start = startSection(BinaryConsts::Section::Table);
-  o << U32LEB(1); // Declare 1 table.
-  o << S32LEB(BinaryConsts::EncodedType::funcref);
-  writeResizableLimits(wasm->table.initial,
-                       wasm->table.max,
-                       wasm->table.hasMax(),
-                       /*shared=*/false,
-                       /*is64*/ false);
+  auto num = importInfo->getNumDefinedTables();
+  o << U32LEB(num);
+  ModuleUtils::iterDefinedTables(*wasm, [&](Table* table) {
+    o << S32LEB(BinaryConsts::EncodedType::funcref);
+    writeResizableLimits(table->initial,
+                         table->max,
+                         table->hasMax(),
+                         /*shared=*/false,
+                         /*is64*/ false);
+  });
   finishSection(start);
 }
 
 void WasmBinaryWriter::writeTableElements() {
-  if (!wasm->table.exists || wasm->table.segments.size() == 0) {
+  size_t elemCount = 0;
+  for (auto& table : wasm->tables) {
+    elemCount += table->segments.size();
+  }
+  if (elemCount == 0) {
     return;
   }
+
   BYN_TRACE("== writeTableElements\n");
   auto start = startSection(BinaryConsts::Section::Element);
+  o << U32LEB(elemCount);
+
+  for (auto& table : wasm->tables) {
+    for (auto& segment : table->segments) {
+      Index tableIdx = getTableIndex(table->name);
 
-  o << U32LEB(wasm->table.segments.size());
-  for (auto& segment : wasm->table.segments) {
-    // Table index; 0 in the MVP (and binaryen IR only has 1 table)
-    o << U32LEB(0);
-    writeExpression(segment.offset);
-    o << int8_t(BinaryConsts::End);
-    o << U32LEB(segment.data.size());
-    for (auto name : segment.data) {
-      o << U32LEB(getFunctionIndex(name));
+      // header format
+      if (tableIdx < 1) {
+        o << U32LEB(0x0);
+      } else {
+        o << U32LEB(0x2);
```

Comment:
It would be good to use the enum values here.

Reply:
Thanks I used constants, although using `?:` made it look a bit ugly! I can try several `if` blocks if you prefer. I've seen both styles in `wasm-binary.cpp`.

Reply:
Yes, I think the `if` blocks would be better in this case, but I'm going to merge this as-is because it is so big and I don't want it to get out of sync with upstream. If you wouldn't mind fixing this in a follow-up PR, that would be great :)